multilog.8
.TH multilog 8
.SH NAME
multilog \- reads a sequence of lines from stdin and appends selected lines to
any number of logs.
.SH SYNOPSIS
.B multilog
.I script
.SH DESCRIPTION
.I script
consists of any number of arguments. Each argument specifies one action. The
actions are carried out in order for each line of input. Note that actions may
contain shell metacharacters that need to be quoted when
.B multilog
is run from a shell. 

.B multilog
exits 0 when it sees the end of stdin. If stdin has a partial final line then
.B multilog
inserts a final newline. 

.B multilog
writes a message to stderr and exits 111, without reading any input, if it
runs out of memory or if another
.B multilog
process is writing to one of the same automatically rotated logs.

If
.B multilog
has trouble writing to disk after it starts reading input, it writes a message
to stderr, pauses, and tries again, without losing any data. Note that this
may block any program feeding input to
.BR multilog .

If
.B multilog
receives a TERM signal, it will read and process data until the next newline,
and then exit, leaving stdin at the first byte of data it has not processed.
.SH SELECTING LINES
Each line is initially selected. The following actions change this:
.TP
.BI - pattern
deselects the line if
.I pattern
matches the line.
.TP
.BI + pattern
selects the line if
.I pattern
matches the line.
.TP
.B F
changes the pattern match mode to use fnmatch for all subsequent patterns.
.TP
.B S
changes the pattern match mode back to use simple patterns (the default).
.PP
.SS SIMPLE PATTERNS
.I pattern
is a string of stars and non-stars. It matches any concatenation of strings
matched by all the stars and non-stars in the same order. A non-star matches
itself. A star before the end of
.I pattern
matches any string that does not include the next character in
.IR pattern .
A star at the end of
.I pattern
matches any string. 

For example, the action 

  +hello

selects hello. It does not select hello world. 

The action 

  -named[*]: Cleaned cache *

deselects named[135]: Cleaned cache of 3121 RRs. The first star matches any
string that does not include a right bracket. 

The action 

  -*

deselects every line. 

To save memory,
.B multilog
actually checks pattern against only the first 1000 characters of each line.
.SS FNMATCH PATTERNS
When fnmatch mode is selected,
.B multilog
uses the system's
.B fnmatch
function to match
.IR pattern .
This mode is less efficient than the simple patterns, but more flexable
in the types of patterns it allows.

In particular,
.TP
.I ?
(not between brackets) matches any single character,
.TP
.I *
(not between brackets) matches any string, including the empty string,
.TP
.I [...]
matches a single character, and
.TP
\fI\\?\fR, \fI\\*\fR, and \fI\\[
match
.IR ? ,
.IR * ,
.I [
exactly, respectively.
.P
The expression
.I [...]
matches any single character enclosed by the brackets.
If the first character within the brackets is
.IR ! ,
the expression matches any character except those enclosed by the brackets.
Two characters within the brackets separated by
.I -
denote a range.
.I [A-F]
is equivalent to
.IR [ABCDEF] .
The characters
.IR ? ,
.IR * ,
and
.I [
lose their special meaning within the brackets.

See the fnmatch(3) or glob(7) man pages for more details.
.SH ALERTS
The action 
.TP
.B e
prints (the first 200 bytes of) each selected line to stderr.
.SH STATUS FILES
The action 
.TP
.BI = file
replaces the contents of
.I file
with (the first 1000 bytes of) each selected line, padded with newlines to
1001 bytes. There is no protection of
.I file
against power outages. 

For example, the sequence of actions 

     -*
     +STAT*
     =log/status

maintains log/status as a copy of the most recent line starting with STAT. 
.SH TIMESTAMPING
The action 
.TP
.B t
inserts an @, a precise timestamp, and a space in front of each line, using
the same format as
.BR tai64n (8).
.TP
.B T
inserts an accustamp-style timestamp and a space in front of each line,
formatted as
.IB S . M
where
.I S
is the number of seconds since the UNIX epoch (midnight January 1
1970-01-01 UTC) and
.I M
is the number of microseconds.
.PP
These are required to be the first action.
.PP
Patterns apply to the line after the timestamp is inserted. For example, if

  multilog t '-*' '+* fatal: *' ./main

reads the line

  fatal: out of memory

then it will log a line such as

  @400000003b4a39c23294b13c fatal: out of memory

with the first * matching the timestamp.

You can use
.BR tai64nlocal (8)
to convert these timestamps to human-readable form.
.SH AUTOMATICALLY ROTATED LOGS
If
.I dir
starts with a dot or slash then the action 
.TP
.I dir
appends each selected line to a log named
.IR dir .
If
.I dir
does not exist,
.B multilog
creates it. 
.PP
Do not attempt to write to one log from two simultaneous
.B multilog
processes, or two actions in one process.

The log format is as follows.
.I dir
is a directory containing some number of old log files, a log file named
.IR current ,
and other files for
.B multilog
to keep track of its actions. Each old log file has a name beginning with @,
continuing with a precise timestamp showing when the file was finished, and
ending with one of the following codes:
.TP
.B .s
This file is completely processed and safely written to disk.  This code
may be changed as described below.
.TP
.B .u
This file was being created at the moment of an outage. It may have been
truncated and has not been processed. 
.PP
Beware that NFS, async filesystems, and softupdates filesystems may discard
files that were not safely written to disk before an outage.

While
.B multilog
is running,
.I current
has mode 644. If
.B multilog
sees the end of stdin, it writes
.I current
safely to disk, and sets the mode of
.I current
to 744. When it restarts, it sets the mode of
.I current
back to 644 and continues writing new lines. 

When
.B multilog
decides that
.I current
is big enough, it writes
.I current
safely to disk, sets the mode of
.I current
to 744, and renames
.I current
as an old log file.
.TP
.BI s size
sets the maximum file size for subsequent
.I dir
actions.
.B multilog
will decide that
.I current
is big enough if
.I current
has
.I size
bytes.
.RB ( multilog
will also decide that
.I current
is big enough if it sees a newline within 2000 bytes of the maximum file size;
it tries to finish log files at line boundaries.)
.I size
must be between 4096 and 2147483647. The default maximum file size is 99999.

In versions 0.75 and above: If
.B multilog
receives an ALRM signal, it immediately decides that
.I current
is big enough, if
.I current
is nonempty.
.TP
.BI n num
sets the number of log files for subsequent
.I dir
actions. After renaming
.IR current ,
if
.B multilog
sees
.I num
or more old log files, it removes the old log file with the smallest
timestamp.
.I num
must be at least 2. The default number of log files is 10.
.TP
.BI ! processor
sets a
.I processor
for subsequent
.I dir
actions.
.B multilog
will feed
.I current
through
.I processor
and save the output as an old log file instead of
.IR current .
.B multilog
will also save any output that
.I processor
writes to descriptor 5, and make that output readable on descriptor 4 when it
runs
.I processor
on the next log file. For reliability,
.I processor
must exit nonzero if it has any trouble creating its output;
.B multilog
will then run it again. Note that running
.I processor
may block any program feeding input to
.BR multilog .
.TP
.BI w code
Changes the "safely written" code from
.B .s
as described above to
.BI . code
for subsequent
.I dir
actions.  This is useful when using a
.I processor
that compresses or otherwise translates the completed log file into a
different file format.
.SH SEE ALSO
fnmatch(3),
glob(7),
envdir(8),
envini(8),
envuidgid(8),
fghack(8),  
pgrphack(8),
readproctitle(8),
setlock(8),
setuidgid(8),
setuser(8),
softlimit(8),
supervise(8),
svc(8),
svok(8),
svscan(8),
svscanboot(8),
svstat(8),
tai64n(8),
tai64nlocal(8)


CHANGES
Added new program setuser to setuid, setenv, plus chdir before exec.

Added new program envini to read environment variables from an INI file.

Added support for adding a prefix in envdir.

Added "once" flag output to svstat.

Fixed sleeper to do a single atomic write.

Added support for resetting limits in softlimit.

Fixed initial supervise "down" status.

Changes in daemontools-encore-1.10
==================================

Added support for fnmatch patterns in multilog.

Increased the multilog file size limit to 2GB.
(Thanks to SATOH Fumiyasu)

Fixed missing separator between system error texts.

Major speed ups in multilog, tai64n, and tai64nlocal.

Added support for signal propagation and disabling setsid behavior in
supervise.  (Thanks to Alan Grow)

Changes in daemontools-encore-1.06
==================================

Added options to svstat to control checking the log script service.

Added -l option to svc to explicitly control the main service.

Fixed race issuing svc commands to log and then main service.

Changed the "safely written" option for multilog to "w".

Fixed daemontools leaking the status file descriptor to its children.

Fixed svc sending commands to both the log and main service.

Changes in daemontools-encore-1.05
==================================

Added options to svup to control checking the log script service.

Eliminated some compiler warnings.

Changes in daemontools-encore-1.04
==================================

Fixed bug in svscan preventing it from starting services with a "log"
script instead of directory.

Changes in daemontools-encore-1.03
==================================

Added -s option to setuidgid to set the supplemental GIDs.
(Thanks to SATOH Fumiyasu)

Changes in daemontools-encore-1.02
==================================

If a supervise is managing both a main and log service, issuing
"svc -dx" on the service will cause the log to stop after the main
service exits.

Changes between daemontools-0.76 and daemontools-encore
=======================================================

Added support to supervise/svc for sending SIGQUIT/USR1/USR2/WINCH.
(original patch by Wayne Marshall, based on patch by Scott Gifford)

Modified supervise to automatically put the forked service into its own
process group to facilitate sending signals to the group.

Added support to supervise/svc for sending signals to the group.

Added support to softlimit for reading limits from environment variables.

Modified supervise to run "start" once before executing "run" to allow
for single-shot initialization, and to run "stop" after "run" exits for
the last time.  The presence of "run" is also optional.

Modified supervise to stop trying to restart the service if it exits 100.
(semantics borrowed from qmail control files)

Added support to multilog for "accustamp" time stamps.
(derived from patch by Matthias Andree)

Added support to supervise for storing the status files in an alternate
directory specified by $SUPERVISEDIR.

Modified multilog to support specifying an alternate "safely written"
filename extension (useful for compressed logs).

Added extended state information to the supervise/status file and svstat.

Added support to supervise for a "notify" script, executed whenever a
subprocess is started or exits.  If present, it is passed the script
name, "start" or "exit", the PID, and the exit code.

Added support for piping to "log" within supervise.  This requires "log"
to be an executable file (instead of a directory as for svscan).

Modified supervise to overwrite status instead of creating a new file.
This allows the "supervise" subdirectory to be non-writeable.

Added a "svup" tool to determine if a service is running.

Modified svscan to optionally redirect all output through a service.
(adapted from http://www.gluelogic.com/code/daemontools/daemontools-0.76-gps-logging-0.01.patch)
Note that this feature works differently from the GlueLogic
modification.  See the svscan(8) man page for details.


multilog.c
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "direntry.h"
#include "alloc.h"
#include "stralloc.h"
#include "buffer.h"
#include "strerr.h"
#include "error.h"
#include "open.h"
#include "lock.h"
#include "scan.h"
#include "str.h"
#include "byte.h"
#include "seek.h"
#include "timestamp.h"
#include "wait.h"
#include "closeonexec.h"
#include "env.h"
#include "fd.h"
#include "sig.h"
#include "match.h"
#include "deepsleep.h"

#define MAXLINE 1000

extern int rename(const char *,const char *);

#define FATAL "multilog: fatal: "
#define WARNING "multilog: warning: "

void pause3(const char *s1,const char *s2,const char *s3)
{
  strerr_warn4sys(WARNING,s1,s2,s3);
  deepsleep(5);
}

void pause5(const char *s1,const char *s2,const char *s3,const char *s4,const char *s5)
{
  strerr_warn6sys(WARNING,s1,s2,s3,s4,s5);
  deepsleep(5);
}

int fdstartdir;

int *f;

void f_init(char **script)
{
  int i;
  int fd;

  for (i = 0;script[i];++i)
    ;

  f = (int *) alloc(i * sizeof(*f));
  if (!f) strerr_die2x(111,FATAL,"out of memory");

  for (i = 0;script[i];++i) {
    fd = -1;
    if (script[i][0] == '=') {
      if (fchdir(fdstartdir) == -1)
        strerr_die2sys(111,FATAL,"unable to switch to starting directory");
      fd = open_append(script[i] + 1);
      if (fd == -1)
        strerr_die3sys(111,FATAL,"unable to create ",script[i] + 1);
      close(fd);
      fd = open_write(script[i] + 1);
      if (fd == -1)
        strerr_die3sys(111,FATAL,"unable to write ",script[i] + 1);
      closeonexec(fd);
    }
    f[i] = fd;
  }
}

struct cyclog {
  char buf[512];
  buffer ss;
  int fdcurrent;
  unsigned long bytes;
  unsigned long num;
  unsigned long size;
  const char *processor;
  const char *code_finished;
  const char *dir;
  int fddir;
  int fdlock;
  int flagselected;
} *c;
int cnum;

stralloc fn = {0};

int filesfit(struct cyclog *d)
{
  DIR *dir;
  direntry *x;
  int count;
  int i;

  dir = opendir(".");
  if (!dir) return -1;

  fn.s[0] = '@';
  fn.s[1] = 'z';
  fn.s[2] = 0;

  count = 0;
  for (;;) {
    errno = 0;
    x = readdir(dir);
    if (!x) break;
    if (x->d_name[0] == '@')
      if (str_len(x->d_name) >= 25) {
        ++count;
        if (str_diff(x->d_name,fn.s) < 0) {
          for (i = 0;i < 25;++i)
            fn.s[i] = x->d_name[i];
          fn.s[25] = 0;
        }
      }
  }
  if (errno) { closedir(dir); return -1; }
  closedir(dir);

  if (count < d->num) return 1;

  dir = opendir(".");
  if (!dir) return -1;

  for (;;) {
    errno = 0;
    x = readdir(dir);
    if (!x) break;
    if (x->d_name[0] == '@')
      if (str_len(x->d_name) >= 25)
        if (str_start(x->d_name,fn.s)) {
	  unlink(x->d_name);
	  break;
        }
  }
  if (errno) { closedir(dir); return -1; }
  closedir(dir);
  return 0;
}

void finish(struct cyclog *d,const char *file,const char *code)
{
  struct stat st;
  int fnlen;

  for (;;) {
    if (stat(file,&st) == 0) break;
    if (errno == error_noent) return;
    pause5("unable to stat ",d->dir,"/",file,", pausing");
  }

  if (st.st_nlink == 1)
    for (;;) {
      fnlen = fmt_tai64nstamp(fn.s);
      fn.s[fnlen++] = '.';
      do {
	fn.s[fnlen++] = *code;
      } while (*code++ != 0);

      if (link(file,fn.s) == 0) break;
      pause5("unable to link to ",d->dir,"/",fn.s,", pausing");
    }

  while (unlink(file) == -1)
    pause5("unable to remove ",d->dir,"/",file,", pausing");

  for (;;)
    switch(filesfit(d)) {
      case 1: return;
      case -1: pause3("unable to read ",d->dir,", pausing");
    }
}

void startprocessor(struct cyclog *d)
{
  const char *args[4];
  int fd;

  sig_uncatch(sig_term);
  sig_uncatch(sig_alarm);
  sig_unblock(sig_term);
  sig_unblock(sig_alarm);

  fd = open_read("previous");
  if (fd == -1) return;
  if (fd_move(0,fd) == -1) return;
  fd = open_trunc("processed");
  if (fd == -1) return;
  if (fd_move(1,fd) == -1) return;
  fd = open_read("state");
  if (fd == -1) return;
  if (fd_move(4,fd) == -1) return;
  fd = open_trunc("newstate");
  if (fd == -1) return;
  if (fd_move(5,fd) == -1) return;

  args[0] = "sh";
  args[1] = "-c";
  args[2] = d->processor;
  args[3] = 0;
  execve("/bin/sh",(char*const*)args,environ);
}

void fullcurrent(struct cyclog *d)
{
  int fd;
  int pid;
  int wstat;

  while (fchdir(d->fddir) == -1)
    pause3("unable to switch to ",d->dir,", pausing");

  while (fsync(d->fdcurrent) == -1)
    pause3("unable to write ",d->dir,"/current to disk, pausing");
  close(d->fdcurrent);

  while (rename("current","previous") == -1)
    pause3("unable to rename current to previous in directory ",d->dir,", pausing");
  while ((d->fdcurrent = open_append("current")) == -1)
    pause3("unable to create ",d->dir,"/current, pausing");
  closeonexec(d->fdcurrent);
  d->bytes = 0;
  while (fchmod(d->fdcurrent,0644) == -1)
    pause3("unable to set mode of ",d->dir,"/current, pausing");

  while (chmod("previous",0744) == -1)
    pause3("unable to set mode of ",d->dir,"/previous, pausing");

  if (!d->processor)
    finish(d,"previous",d->code_finished);
  else {
    for (;;) {
      while ((pid = fork()) == -1)
        pause3("unable to fork for processor in ",d->dir,", pausing");
      if (!pid) {
        startprocessor(d);
        strerr_die3sys(111,FATAL,"unable to run ",d->processor);
      }
      if (wait_pid(&wstat,pid) == -1)
        pause3("wait failed for processor in ",d->dir,", pausing");
      else if (wait_crashed(wstat))
        pause3("processor crashed in ",d->dir,", pausing");
      else if (!wait_exitcode(wstat))
        break;
      strerr_warn4(WARNING,"processor failed in ",d->dir,", pausing",0);
      deepsleep(5);
    }

    while ((fd = open_append("processed")) == -1)
      pause3("unable to create ",d->dir,"/processed, pausing");
    while (fsync(fd) == -1)
      pause3("unable to write ",d->dir,"/processed to disk, pausing");
    while (fchmod(fd,0744) == -1)
      pause3("unable to set mode of ",d->dir,"/processed, pausing");
    close(fd);

    while ((fd = open_append("newstate")) == -1)
      pause3("unable to create ",d->dir,"/newstate, pausing");
    while (fsync(fd) == -1)
      pause3("unable to write ",d->dir,"/newstate to disk, pausing");
    close(fd);

    while (unlink("previous") == -1)
      pause3("unable to remove ",d->dir,"/previous, pausing");
    while (rename("newstate","state") == -1)
      pause3("unable to rename newstate to state in directory ",d->dir,", pausing");
    finish(d,"processed",d->code_finished);
  }
}

int c_write(int pos,char *buf,int len)
{
  struct cyclog *d;
  int w;

  d = c + pos;

  if (d->bytes >= d->size)
    fullcurrent(d);

  if (len >= d->size - d->bytes)
    len = d->size - d->bytes;

  if (d->bytes + len >= d->size - 2000) {
    w = byte_rchr(buf,len,'\n');
    if (w < len)
      len = w + 1;
  }

  for (;;) {
    w = write(d->fdcurrent,buf,len);
    if (w > 0) break;
    pause3("unable to write to ",d->dir,"/current, pausing");
  }

  d->bytes += w;
  if (d->bytes >= d->size - 2000)
    if (buf[w - 1] == '\n')
      fullcurrent(d);

  return w;
}

void restart(struct cyclog *d)
{
  struct stat st;
  int fd;
  int flagprocessed;

  if (fchdir(fdstartdir) == -1)
    strerr_die2sys(111,FATAL,"unable to switch to starting directory");

  mkdir(d->dir,0700);
  d->fddir = open_read(d->dir);
  if ((d->fddir == -1) || (fchdir(d->fddir) == -1))
    strerr_die3sys(111,FATAL,"unable to open directory ",d->dir);
  closeonexec(d->fddir);

  d->fdlock = open_append("lock");
  if ((d->fdlock == -1) || (lock_exnb(d->fdlock) == -1))
    strerr_die3sys(111,FATAL,"unable to lock directory ",d->dir);
  closeonexec(d->fdlock);

  if (stat("current",&st) == -1) {
    if (errno != error_noent)
      strerr_die4sys(111,FATAL,"unable to stat ",d->dir,"/current");
  }
  else
    if (st.st_mode & 0100) {
      fd = open_append("current");
      if (fd == -1)
        strerr_die4sys(111,FATAL,"unable to append to ",d->dir,"/current");
      if (fchmod(fd,0644) == -1)
        strerr_die4sys(111,FATAL,"unable to set mode of ",d->dir,"/current");
      closeonexec(fd);
      d->fdcurrent = fd;
      d->bytes = st.st_size;
      return;
    }

  unlink("state");
  unlink("newstate");

  flagprocessed = 0;
  if (stat("processed",&st) == -1) {
    if (errno != error_noent)
      strerr_die4sys(111,FATAL,"unable to stat ",d->dir,"/processed");
  }
  else if (st.st_mode & 0100)
    flagprocessed = 1;

  if (flagprocessed) {
    unlink("previous");
    finish(d,"processed",d->code_finished);
  }
  else {
    unlink("processed");
    finish(d,"previous","u");
  }

  finish(d,"current","u");

  fd = open_trunc("state");
  if (fd == -1)
    strerr_die4sys(111,FATAL,"unable to write to ",d->dir,"/state");
  close(fd);
  fd = open_append("current");
  if (fd == -1)
    strerr_die4sys(111,FATAL,"unable to write to ",d->dir,"/current");
  if (fchmod(fd,0644) == -1)
    strerr_die4sys(111,FATAL,"unable to set mode of ",d->dir,"/current");
  closeonexec(fd);
  d->fdcurrent = fd;
  d->bytes = 0;
}

void c_init(char **script)
{
  int i;
  struct cyclog *d;
  const char *processor;
  unsigned long num;
  unsigned long size;
  const char *code_finished = "s";

  cnum = 0;
  for (i = 0;script[i];++i)
    if ((script[i][0] == '.') || (script[i][0] == '/'))
      ++cnum;

  c = (struct cyclog *) alloc(cnum * sizeof(*c));
  if (!c) strerr_die2x(111,FATAL,"out of memory");

  d = c;
  processor = 0;
  num = 10;
  size = 99999;

  for (i = 0;script[i];++i)
    if (script[i][0] == 's') {
      scan_ulong(script[i] + 1,&size);
      if (size < 4096) size = 4096;
      if (size > 2147483647) size = 2147483647;
    }
    else if (script[i][0] == 'n') {
      scan_ulong(script[i] + 1,&num);
      if (num < 2) num = 2;
    }
    else if (script[i][0] == '!') {
      processor = script[i] + 1;
    }
    else if (script[i][0] == 'w') {
      code_finished = script[i] + 1;
      if (!stralloc_ready(&fn,str_len(code_finished)+TIMESTAMP+1))
	strerr_die2sys(111,FATAL,"unable to allocate memory");
    }
    else if ((script[i][0] == '.') || (script[i][0] == '/')) {
      d->num = num;
      d->size = size;
      d->processor = processor;
      d->code_finished = code_finished;
      d->dir = script[i];
      buffer_init(&d->ss,c_write,d - c,d->buf,sizeof d->buf);
      restart(d);
      ++d;
    }
}

void c_quit(void)
{
  int j;

  for (j = 0;j < cnum;++j) {
    buffer_flush(&c[j].ss);
    while (fsync(c[j].fdcurrent) == -1)
      pause3("unable to write ",c[j].dir,"/current to disk, pausing");
    while (fchmod(c[j].fdcurrent,0744) == -1)
      pause3("unable to set mode of ",c[j].dir,"/current, pausing");
  }
}

int flagexitasap = 0;
int flagforcerotate = 0;
int flagnewline = 1;

void exitasap(void)
{
  flagexitasap = 1;
}

void forcerotate(void)
{
  flagforcerotate = 1;
}

int flushread(int fd,char *buf,int len)
{
  int j;

  for (j = 0;j < cnum;++j)
    buffer_flush(&c[j].ss);
  if (flagforcerotate) {
    for (j = 0;j < cnum;++j)
      if (c[j].bytes > 0)
	fullcurrent(&c[j]);
    flagforcerotate = 0;
  }

  if (!len) return 0;
  if (flagexitasap) {
    if (flagnewline) return 0;
    len = 1;
  }

  sig_unblock(sig_term);
  sig_unblock(sig_alarm);

  len = read(fd,buf,len);

  sig_block(sig_term);
  sig_block(sig_alarm);

  if (len <= 0) return len;
  flagnewline = (buf[len - 1] == '\n');
  return len;
}

char inbuf[1024];
buffer ssin = BUFFER_INIT(flushread,0,inbuf,sizeof inbuf);

char line[MAXLINE+1];

void doit(char **script)
{
  int flageof;
  char ch;
  int j;
  int i;
  const char *action;
  int flagselected;
  int flagtimestamp;
  unsigned int linelen; /* 0 <= linelen <= MAXLINE */
  unsigned int oldlinelen = sizeof line;

  flagtimestamp = 0;
  if (script[0])
    if (script[0][0] == 't' || script[0][0] == 'T')
      flagtimestamp = script[0][0];

  for (i = 0;i <= MAXLINE;++i) line[i] = '\n';

  flageof = 0;
  for (;;) {
    linelen = 0;

    if (buffer_feed(&ssin) <= 0)
      return;
    if (flagtimestamp) {
      linelen = (flagtimestamp == 't')
	? fmt_tai64nstamp(line)
	: fmt_accustamp(line);
      line[linelen++] = ' ';
    }
    if (buffer_gets(&ssin,line+linelen,MAXLINE-linelen,'\n',&linelen) < 0)
      flageof = 1;

    for (i = linelen;i < oldlinelen;++i) line[i] = '\n';

    flagselected = 1;
    j = 0;
    for (i = 0;(action = script[i]) != 0;++i)
      switch(*action) {
        case 'F':
	  match = match_fnmatch;
	  break;
        case 'S':
	  match = match_simple;
	  break;
        case '+':
          if (!flagselected)
            if (match(action + 1,line,linelen))
              flagselected = 1;
          break;
        case '-':
          if (flagselected)
            if (match(action + 1,line,linelen))
              flagselected = 0;
          break;
        case 'e':
          if (flagselected) {
            if (linelen > 200) {
              buffer_put(buffer_2,line,200);
              buffer_puts(buffer_2,"...\n");
            }
            else {
              buffer_put(buffer_2,line,linelen);
              buffer_puts(buffer_2,"\n");
            }
            buffer_flush(buffer_2);
          }
          break;
        case '=':
          if (flagselected)
            for (;;) {
              while (seek_begin(f[i]) == -1)
                pause3("unable to move to beginning of ",action + 1,", pausing");
              if (write(f[i],line,MAXLINE+1) == MAXLINE+1)
                break;
              pause3("unable to write ",action + 1,", pausing");
            }
          break;
        case '.':
        case '/':
          c[j].flagselected = flagselected;
          ++j;
          break;
      }

    for (j = 0;j < cnum;++j)
      if (c[j].flagselected)
        buffer_put(&c[j].ss,line,linelen);
        
    while (linelen == MAXLINE) {
      linelen = 0;
      if (buffer_gets(&ssin,line,MAXLINE,'\n',&linelen) < 0) {
	flageof = 1;
	break;
      }
      if (linelen == 0)
	break;
      for (j = 0;j < cnum;++j)
	if (c[j].flagselected)
	  buffer_put(&c[j].ss,line,linelen);
    }

    for (j = 0;j < cnum;++j)
      if (c[j].flagselected) {
	ch = '\n';
        buffer_PUTC(&c[j].ss,ch);
      }

    if (flageof) return;

    oldlinelen = linelen;
  }
}

int main(int argc,char **argv)
{
  umask(022);

  if (!stralloc_ready(&fn,40))
    strerr_die2sys(111,FATAL,"unable to allocate memory");

  fdstartdir = open_read(".");
  if (fdstartdir == -1)
    strerr_die2sys(111,FATAL,"unable to switch to current directory");
  closeonexec(fdstartdir);

  sig_block(sig_term);
  sig_block(sig_alarm);
  sig_catch(sig_term,exitasap);
  sig_catch(sig_alarm,forcerotate);

  ++argv;
  f_init(argv);
  c_init(argv);
  doit(argv);
  c_quit();
  _exit(0);
}
